File: catalog/backends/java-spring/patterns/KafkaConsumerService.java
package com.example.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.stereotype.Service;

// Pattern: DSM — Kafka consumer with optional lag simulation
// Adapt: replace processing logic with domain-specific handling
@Service
public class KafkaConsumerService {

    private static final Logger log = LoggerFactory.getLogger(KafkaConsumerService.class);
    private static final long DELAY = Long.parseLong(
            System.getenv().getOrDefault("KAFKA_CONSUMER_DELAY_MS", "0"));

    @KafkaListener(topics = "demo-events", groupId = "demo-consumer-group")
    public void consume(String message) throws InterruptedException {
        log.info("consumed event: {}", message);
        if (DELAY > 0) Thread.sleep(DELAY);
    }
}

File: catalog/backends/dotnet-aspnetcore/patterns/KafkaConsumerService.cs
using Confluent.Kafka;
using Microsoft.Extensions.Logging;

// Pattern: DSM — Kafka consumer with optional lag simulation
// Adapt: replace processing logic with domain-specific handling
public class KafkaConsumerService
{
    private static readonly int DelayMs = int.TryParse(
        Environment.GetEnvironmentVariable("KAFKA_CONSUMER_DELAY_MS"), out var d) ? d : 0;

    public static void StartConsumer(ILogger logger)
    {
        var config = new ConsumerConfig
        {
            BootstrapServers = Environment.GetEnvironmentVariable("KAFKA_BOOTSTRAP_SERVERS") ?? "kafka:9092",
            GroupId = "demo-consumer-group",
            AutoOffsetReset = AutoOffsetReset.Latest,
        };
        using var consumer = new ConsumerBuilder<string, string>(config).Build();
        consumer.Subscribe("demo-events");
        while (true)
        {
            var result = consumer.Consume(TimeSpan.FromSeconds(1));
            if (result == null) continue;
            logger.LogInformation("consumed event: {Value}", result.Message.Value);
            if (DelayMs > 0) Thread.Sleep(DelayMs);
        }
    }
}
